### PR TITLE
eth/types: add `Num` superclass and `Hex`, `Dec`, `Bin` descendants

### PR DESCRIPTION
**Pull request by mculp**
Porting the Num type from https://github.com/q9f/secp256k1.cr/blob/master/src/secp256k1/num.cr

### CODE REVIEW

File: lib/eth/types/faceted_numeric.rb [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    #
+    # @see {https://github.com/q9f/secp256k1.cr/blob/master/src/secp256k1/num.cr} Num from secp256k1.cr
+    # @see #{Eth::Util} is used heavily throughout
+    # @example Pass in a 0x-prefixed hex string, and coerce to keccak256 binary
+    #   FacetedNumeric.new('0xDEADBeeF').to_binary
+    #   FacetedNumeric['0xDEADBeeF'].to_keccak256
+    # @raise [ArgumentError] raised if input is a String but
+    #   isn't hex or bytes, or if input is an Array but not
+    #   a list of ascii-8bit numbers
+    class FacetedNumeric < Numeric
```
Comment (q9f) [inline lib/eth/types/faceted_numeric.rb:35]: It is to be expected that this datatype is to be used across the entire library, I would encourage you to pick a short name for the class, e.g., `::Num`, and use the documentation to elaborate on the nature of this faceted numeric type.

```suggestion
    class Num < Numeric
```

File: lib/eth/types/faceted_numeric.rb [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+      include Util
```
Comment (q9f) [inline lib/eth/types/faceted_numeric.rb:36]: Ideally, the goal should be to get rid of anything that works with numerics in `::Util` and replace this with native functions in the `::Num` type.

```suggestion
```

File: lib/eth/types/faceted_numeric.rb [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+      # raw input
+      attr_reader :input
+
+      # value in base 10
+      attr_reader :integer
+
+      # binary (keccak256)
+      # @see {Eth::Util.keccak256}
+      attr_reader :binary
```
Comment (q9f) [inline lib/eth/types/faceted_numeric.rb:46]: I believe, keccak is just one of many use-cases for binary numerics.

File: lib/eth/types/faceted_numeric.rb [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+      # hexadecimal
+      attr_reader :hex
+
+      # 0x-prefixed hexadecimal
+      attr_reader :prefixed_hex
```
Comment (q9f) [inline lib/eth/types/faceted_numeric.rb:52]: This should be a function, taking the functionality from the `Util::` module.

File: lib/eth/types/faceted_numeric.rb [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+      # zero-padded hexadecimal
+      attr_reader :zpadded_hex
```
Comment (q9f) [inline lib/eth/types/faceted_numeric.rb:55]: This should be a function, taking the functionality from the `Util::` module.

File: lib/eth/types/faceted_numeric.rb [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+
+      # zero-padded binary
+      attr_reader :zpadded_binary
```
Comment (q9f) [inline lib/eth/types/faceted_numeric.rb:58]: This should be a function, taking the functionality from the `Util::` module.

File: lib/eth/types/faceted_numeric.rb [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+
+      def to_hex
+        hex 
+      end
+
+      def to_binary
+        binary
+      end
+
+      def to_keccak256
```
Comment (q9f) [inline lib/eth/types/faceted_numeric.rb:107]: This should not be part of the numeric type.

File: lib/eth/types/faceted_numeric.rb [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+        keccak256(to_prefixed_hex)
+      end
+
+      def to_prefixed_hex
+        @prefixed_hex ||= prefix_hex(hex)
+      end
+
+      def to_zpadded_hex
```
Comment (q9f) [inline lib/eth/types/faceted_numeric.rb:115]: Ideally, you want to allow specifying how many zeros you want to pad.

**Review by q9f — COMMENTED** [review]
Thank you. I already had some opinions and left some comments.

File: lib/eth/types/faceted_numeric.rb (hunk 3 above)
Comment (mculp) [reply lib/eth/types/faceted_numeric.rb:46]: yeah, I figured out keccak and hex_to_bin weren't providing the same output after writing this doc

File: lib/eth/types/faceted_numeric.rb (hunk 7 above)
Comment (mculp) [reply lib/eth/types/faceted_numeric.rb:107]: This topic was brought up on @josh's PR where he was converting from keccak to prefixed hex.

I see what you mean, but I don't see how this class will help that  particular use case much. They're almost the same length.

`Util.prefix_hex(Util.bin_to_hex(Util.keccak256(sig)))`

`Num.new(Eth::Util.keccak256(sig)).to_prefixed_hex` 

Alternatively, I had planned to add `[]` or `()` syntax for instantiation, and we can remove the `Eth::` namespace.

`Num[Util.keccak256(sig)].to_prefixed_hex`

What about naming (or at least aliasing) the `to_prefixed_hex` method `to_0x` or `to_0x_hex`? There's a lot of precedence for short `to_` methods, like `to_i`, `to_f`, `to_d`, etc

`Num[Util.keccak256(sig)].to_0x`

File: lib/eth/types/faceted_numeric.rb (hunk 4 above)
Comment (mculp) [reply lib/eth/types/faceted_numeric.rb:52]: Yep! the `to_` methods are implemented below. This is for memoization. The `to_prefixed_hex` method memoizes the value as `@prefixed_hex`.

I suppose we don't need a getter and we could keep that instance variable value private.

File: lib/eth/types/faceted_numeric.rb (hunk 1 above)
Comment (mculp) [reply lib/eth/types/faceted_numeric.rb:35]: You think it should be a top level class? Get rid of `Types` namespace?

File: lib/eth/types/faceted_numeric.rb (hunk 2 above)
Comment (mculp) [reply lib/eth/types/faceted_numeric.rb:36]: cool! sounds good.

File: lib/eth/types/faceted_numeric.rb (hunk 7 above)
Comment (q9f) [reply lib/eth/types/faceted_numeric.rb:107]: No, the keccak() function will return a Num and all you do is call .to_prefixed_hex afterwards.

Let's implement the Num first and then we can start refactoring the library to use it everywhere.

File: lib/eth/types/faceted_numeric.rb (hunk 8 above)
Comment (mculp) [reply lib/eth/types/faceted_numeric.rb]: Fixed

**Review by q9f — DISMISSED** [review]
Thank you, Matt.

I'm not very happy with how Ruby handles types. Therefore, I think using refinements should be avoided.

If we want to use the `Num` type throughout the library, we should handle all conversions internally in the `Num` class privately and only expose some getters for the different formats. 

What do you think?

**Comment by mculp** [comment]
> Thank you, Matt.
> 
> I'm not very happy with how Ruby handles types. Therefore, I think using refinements should be avoided.
> 
> If we want to use the `Num` type throughout the library, we should handle all conversions internally in the `Num` class privately and only expose some getters for the different formats.
> 
> What do you think?

Just curious, what do you see as the issue with using refinements? A refinement is scoped to the file in which it is used. It's opt-in, file-by-file. So, as this PR stands, the only class that can see the changes to the `String`/`Integer` classes is `Num`. I was thinking we could add further conversions in the refinement (to_keccak, for example) and other classes could opt-in to using the refinement as they want.

If refinements are out, I think the next-best solution would be to subclass Num w/ `Hex`, `Dec`, `Bin`:

```ruby
class Num
  attr_reader :hex, :dec, :bin
  
  def initialize(input = nil)
    generate_number
  end
  
  def to_bytes
    bin
  end
  
  def to_i
    dec
  end
  
  def to_hex
    hex
  end
end

class Hex < Num
  def initialize(input)
    @hex = input.delete_prefix('0x')
    @dec = hex.to_i(16)
    @bin = hex.scan(/../).map(&:hex).pack('C*').b
  end
end

Hex['0xasdfa'].to_bytes
```

I was going back and forth with this or having subclasses of Num; we can't do a straight port from Crystal because you can't define multiple constructors in Ruby. I did see you added a gem to do something like that, but the "multiple constructors" are actually just builder/factory methods if I understand it correctly. i.e. we'd have to do something like this: `Num.from_hex('0xafsdfasdf')` (we want something short and sweet) and there'd be a bunch of nested branching logic, [which](https://medium.com/stantmob/avoiding-conditionals-a2cc5cf38e93) [is](https://github.com/rubocop/ruby-style-guide#no-nested-conditionals) [typically](https://www.rubyguides.com/2017/04/stop-using-case-statements-in-ruby/) [avoided](http://haacked.com/archive/2013/11/08/death-to-the-if-statement.aspx/) if possible in Ruby

Now that I've looked at the code I typed above, using inheritance would probably be the most idiomatic solution out of the 3.

What do you think?

(btw, I'm not trying to be difficult. I'm just trying to solve this the simplest and most idiomatic way)

**Comment by q9f** [comment]
> Now that I've looked at the code I typed above, using inheritance would probably be the most idiomatic solution out of the 3.

Yes, I love it. I'm not too concerned about the implications of the refinements, it's more about the style. 

Sorry for being so pedantic. I think, your last suggestion most simple and clean approach.

**Comment by mculp** [comment]
> > Now that I've looked at the code I typed above, using inheritance would probably be the most idiomatic solution out of the 3.
> 
> Yes, I love it. I'm not too concerned about the implications of the refinements, it's more about the style.
> 
> Sorry for being so pedantic. I think, your last suggestion most simple and clean approach.

Awesome! What do you think if the files live in `eth/types/...` but then have their constants hoisted to `Eth` like `Eth::Num = Eth::Types::Num`, which would be shortened to `Num` in practice because everything is within the `Eth` namespace.

**Comment by q9f** [comment]
This is looking good so far, do you need a hand to finish this off? :)

**Comment by mculp** [comment]
hey @q9f so sorry for the delay, I took a month vacation. Will finish this asap.

**Comment by mculp** [comment]
hey @q9f I made some updates, let me know what you think.

**Comment by mculp** [comment]
Is the CI broken? It says it's missing a token

**Comment by q9f** [comment]
Can you run rufo on the code? I will review asap.

You can ignore the missing token issue on CI. It should be temporary.

**Comment by mculp** [comment]
@q9f I updated this with Rufo.

**Comment by q9f** [comment]
one test is failing, please check the CI

**Comment by mculp** [comment]
Going down a rabbit hole with this one test.

@q9f I should've asked this question a long time ago, but I noticed in the Crystal `Num`, it was essentially 3 types: Hex, Decimal, and an array of ASCII chars.

Do we want to do that here instead of this bytestring stuff I'm trying to do? The reason I chose a bytestring instead of an array of bytes is because the Keccak lib dumps a bytestring.

Now:

```ruby
    # @example Instantiate a Hex object
    #   Hex('0xdeadbeef')
    #     => #<Eth::Types::Hex:0x00007f8dcb102330
    #       @input="0xdeadbeef",
    #       @hex="deadbeef",
    #       @bin="\xDE\xAD\xBE\xEF",
    #       @dec=3735928559>
```

Would be much easier to deal with, but may not provide the same utility/helpfulness:

```ruby
    # @example Instantiate a Hex object
    #   Hex('0xdeadbeef')
    #     => #<Eth::Types::Hex:0x00007f8dcb102330
    #       @input="0xdeadbeef",
    #       @hex="deadbeef",
    #       @bin=[51, 55, 51, 53, 57, 50, 56, 53, 53, 57],
    #       @dec=3735928559>
```

What do you think? This PR is starting to feel like a monkey on my back.

**Comment by q9f** [comment]
Let's bench this for now

**Pull request closed without merging**